lib/maths/CBoostedTreeFactory.cc

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <maths/CBoostedTreeFactory.h>
+
+#include <maths/CBayesianOptimisation.h>
+#include <maths/CBoostedTreeImpl.h>
+
+namespace ml {
+namespace maths {
+
+CBoostedTreeFactory::TCBoostedTreeUPtr CBoostedTreeFactory::build() {
+    this->initializeMissingFeatureMasks(*m_Frame);
+    std::tie(m_Tree->m_Impl->m_TrainingRowMasks,
+             m_Tree->m_Impl->m_TestingRowMasks) = this->crossValidationRowMasks();
```

Comment:
I think it would be cleaner if you had a constructor on `CBoostedTree` which takes `CBoostedTreeImpl` and have have `m_Tree` be of type `CBoostedTreeImpl` then you can ditch all the `->m_Impl`s in this code.

Reply:
In this case I would need to expose the `Impl` class in method `incompleteTreeObject`, which I cannot do. IMO, either we have to find another way for memory estimation or we have to live with the `->m_Impl`s :man_shrugging:

Reply:
I kind of wonder if `estimateMemoryUsage` should be on the factory object anyway. To me it makes more sense that one asks the factory how big the object it is being asked to build is going to be than get an incompletely built object which can only be used for this purpose. That way it can call the same method on the implementation itself and this never gets exposed to the calling code. What do you think?

Reply:
I think, factory should have a single responsibility of building a `CBoostedTree` objects. My first intention was to introduce a Proxy for `CBoostedTree`, which can handle memory estimation, but initialized the proper boosted tree implementation if called with `train()`. `incompleteTreeObject` is the best I could come up with at the moment, since it still kind-of creating an object and hence does not violate SRP. But the rest of the process is finished, we may see a more elegant solution.

Reply:
I guess this comes down to whose concern you think memory estimation is. I'm fine to kick this decision to the end and it might be easiest to discuss a bit offline.